core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -2017,18 +2011,23 @@ private[spark] object Utils extends Logging {
       conf: SparkConf,
       serviceName: String = ""): (T, Int) = {
 
+    val maxPort = 65536
+    val failedPorts  = new ArrayBuffer[Int]()
     require(startPort == 0 || (1024 <= startPort && startPort < 65536),
       "startPort should be between 1024 and 65535 (inclusive), or 0 for a random free port.")
-
     val serviceString = if (serviceName.isEmpty) "" else s" '$serviceName'"
     val maxRetries = portMaxRetries(conf)
     for (offset <- 0 to maxRetries) {
       // Do not increment port if startPort is 0, which is treated as a special port
       val tryPort = if (startPort == 0) {
         startPort
-      } else {
-        // If the new port wraps around, do not try a privilege port
-        ((startPort + offset - 1024) % (65536 - 1024)) + 1024
+      }
+      else {
+        // If the new port wraps around, do not try a privilege port or a failedPort
+        while(!failedPorts.contains((startPort + Math.random() * (maxPort - startPort + 1))) &&
```

Comment:
This is an array buffer, so contains is linear time, which is pretty expensive. Also this code is very difficult to understand.